src/shell.rs

```diff
@@ -92,434 +75,310 @@ enum AsyncCommandBehavior {
+            sequence,
+            state.clone(),
+            last_input.take().unwrap(),
+            stdout,
+            environment.clone(),
+          );
+          last_input = Some(stdin);
+          wait_tasks.push(spawned_sequence);
+        }
+        let output_handle = tokio::task::spawn_blocking(|| {
+          last_input.unwrap().pipe_to_sender(stdout).unwrap();
+        });
+        let mut results = futures::future::join_all(wait_tasks).await;
+        output_handle.await.unwrap();
+        let last_result = results.pop().unwrap();
+        let all_handles = results.into_iter().flat_map(|r| r.into_handles());
+        match last_result {
+          ExecuteResult::Exit(code, mut handles) => {
+            handles.extend(all_handles);
+            ExecuteResult::Continue(code, Vec::new(), handles)
+          }
+          ExecuteResult::Continue(code, _, mut handles) => {
+            handles.extend(all_handles);
+            ExecuteResult::Continue(code, Vec::new(), handles)
           }
-          .boxed(),
         }
       }
       Sequence::Subshell(list) => {
-        let (stdout_tx, stdout) = ShellPipe::channel();
-        SpawnedStep {
+        let result = spawn_sequential_list(
+          *list,
+          state.clone(),
+          stdin,
           stdout,
-          task: async move {
-            let step = spawn_sequential_list(
-              *list,
-              state.clone(),
-              stdin,
-              environment,
-              // yield async commands to the parent
-              AsyncCommandBehavior::Yield,
-            );
-            let result = step
-              .wait_with_stdout(move |stdout| async move {
-                stdout.pipe_to_sender(stdout_tx).await;
-              })
-              .await;
-
-            // sub shells do not cause an exit
-            match result {
-              ExecuteResult::Exit(code, handles) => {
-                ExecuteResult::Continue(code, Vec::new(), handles)
-              }
-              ExecuteResult::Continue(_, _, _) => result,
-            }
+          environment,
+          // yield async commands to the parent
+          AsyncCommandBehavior::Yield,
+        )
+        .await;
+
+        // sub shells do not cause an exit
+        match result {
+          ExecuteResult::Exit(code, handles) => {
+            ExecuteResult::Continue(code, Vec::new(), handles)
           }
-          .boxed(),
+          ExecuteResult::Continue(_, _, _) => result,
         }
       }
     }
   }
   .boxed()
 }
 
-async fn spawn_and_wait_sequence(
-  sequence: Sequence,
-  state: ShellState,
-  stdin: ShellPipe,
-  sender: ShellPipeSender,
-  environment: impl Environment,
-) -> ExecuteResult {
-  let step = spawn_sequence(sequence, state, stdin, environment).await;
-  step
-    .wait_with_stdout(move |stdout| async {
-      stdout.pipe_to_sender(sender).await;
-    })
-    .await
-}
-
-async fn start_command(
+fn start_command(
```

Comment:
Why return future instead of making the function `async`? Is it still caused by need to recursive calls?

Reply:
Nope, no need for this. Will do this update in the next pr. Thanks!